compiler/javac/src/org/jetbrains/kotlin/javac/Javac.kt

```diff
@@ -0,0 +1,227 @@
+import com.sun.tools.javac.main.JavaCompiler
+import com.sun.tools.javac.model.JavacElements
+import com.sun.tools.javac.tree.JCTree
+import com.sun.tools.javac.util.Context
+import com.sun.tools.javac.util.Names
+import com.sun.tools.javac.util.Options
+import org.jetbrains.kotlin.cli.common.messages.CompilerMessageLocation
+import org.jetbrains.kotlin.cli.common.messages.CompilerMessageSeverity
+import org.jetbrains.kotlin.cli.common.messages.MessageCollector
+import com.sun.tools.javac.util.List as JavacList
+import org.jetbrains.kotlin.wrappers.symbols.JavacClass
+import org.jetbrains.kotlin.wrappers.symbols.JavacPackage
+import org.jetbrains.kotlin.load.java.structure.JavaClass
+import org.jetbrains.kotlin.name.FqName
+import org.jetbrains.kotlin.name.isSubpackageOf
+import org.jetbrains.kotlin.name.parentOrNull
+import org.jetbrains.kotlin.wrappers.trees.JCClass
+import org.jetbrains.kotlin.wrappers.trees.JCPackage
+import java.io.Closeable
+import java.io.File
+import javax.lang.model.element.TypeElement
+import javax.tools.JavaFileManager
+import javax.tools.JavaFileObject
+import javax.tools.StandardLocation
+
+class Javac(javaFiles: Collection<File>,
+            classPathRoots: List<File>,
+            outDir: File?,
+            private val messageCollector: MessageCollector?,
+            arguments: Array<String>?) : Closeable {
+
+    companion object {
+        fun getInstance(project: Project): Javac = ServiceManager.getService(project, Javac::class.java)
+    }
+
+    val JAVA_LANG_OBJECT by lazy { findClassInSymbols(CommonClassNames.JAVA_LANG_OBJECT) }
+
+    private val context = Context()
+
+    init {
+        messageCollector?.let { JavacLogger.preRegister(context, it) }
+        arguments?.toList()?.let { JavacOptionsMapper.map(Options.instance(context), it) }
+    }
+
+    private val javac = object : JavaCompiler(context) {
+        override fun parseFiles(files: Iterable<JavaFileObject>?) = compilationUnits
+    }
+    private val fileManager = context[JavaFileManager::class.java] as JavacFileManager
+
+    init {
+        fileManager.setLocation(StandardLocation.CLASS_PATH, classPathRoots)
+        outDir?.let {
+            it.mkdirs()
+            fileManager.setLocation(StandardLocation.CLASS_OUTPUT, listOf(it))
+        }
+    }
+
+    private val symbols = Symtab.instance(context)
+    private val trees = JavacTrees.instance(context)
+    private val elements = JavacElements.instance(context)
+    private val fileObjects = fileManager.getJavaFileObjectsFromFiles(javaFiles).toJavacList()
+    private val compilationUnits: JavacList<JCTree.JCCompilationUnit> = fileObjects.map(javac::parse).toJavacList()
+
+    private val javaClasses = compilationUnits
+            .flatMap { unit -> unit.typeDecls
+                    .flatMap { JCClass(it as JCTree.JCClassDecl, trees.getPath(unit, it), this).withInnerClasses() }
+            }
+            .map { it.fqName to it }
+            .toMap()
```

Comment:
There is `keysToMap` in Kotlin project itself & also `associateBy` in stdlib. I'm not sure it's quite important but they may be faster (one internal container instead of two) and definitely shorter.